### PR TITLE
parallel max and min for ATen on CPU

### PR DESCRIPTION
**Pull request by mingfeima**
optimize max and min reduction for ATen CPU path, current code path from TH module runs in sequential on CPU.

### CODE REVIEW

**Comment by fmassa** [comment]
cc @colesbury

**Comment by mingfeima** [comment]
can someone take a look at the build fail:
https://ci.pytorch.org/jenkins/job/caffe2-builds/job/conda2-macos10.13-build/9853/console
doesn't seem to be a compilation issue.

**Comment by SsnL** [comment]
Feel free to ignore that one.

Do you have some benchmarks on this?

**Comment by mingfeima** [comment]
@SsnL sure, i wrote a small benchmark for max, the piece of code reduces from `49ms` to `3.5ms` on Xeon Skylake 8180.

```python
import torch
from time import time

N = 2000
T = 35820
warmups = 100
count = 200

a = torch.randn(N, T)

def test_max():
    for i in range(warmups):
        b, _ = a.max(dim=1)
        tstart = time()
        for i in range(count):
        b, _ = a.max(dim=1)
        tend = time()
    print("max reduction : %f ms" % ((tend-tstart)/count*1000))

test_max()
```

I brought this up because i have been optimizing [OpenNMT-py](https://github.com/OpenNMT/OpenNMT-py), with `max` being paralleled, the total time reduced roughly by 8%.

File: aten/src/ATen/native/cpu/TensorCompareKernel.cpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+namespace at { namespace native { namespace {
+
+template <typename scalar_t>
+bool _isnan(scalar_t val) {
+  return false;
+}
+
+template <>
+bool _isnan(float val) {
+  return std::isnan(val);
```
Comment (colesbury) [inline aten/src/ATen/native/cpu/TensorCompareKernel.cpp]: Can you just use `std::nan()` directly? It works on integer types and GCC and clang optimize it to `false`/`0` on integers at `-O1` or higher.

File: aten/src/ATen/native/cpu/TensorCompareKernel.cpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+template <>
+bool _isnan(double val) {
+  return std::isnan(val);
+}
+
+#define isnan_break(val) \
```
Comment (colesbury) [inline aten/src/ATen/native/cpu/TensorCompareKernel.cpp]: I get a bit confused by the control-flow via macro. Can you replace the usages with the if statement?

**Review by colesbury — COMMENTED** [review]
Nice speed-ups. LGTM with a few code style comments.

I'm working on changing how reductions are implemented and unifying some of the CPU and CUDA code, but it'll probably take a while, so this speed-up is very welcome.

File: aten/src/ATen/native/cpu/TensorCompareKernel.cpp (hunk 1 above)
Comment (mingfeima) [reply aten/src/ATen/native/cpu/TensorCompareKernel.cpp]: sorry, i don't quite follow here. Could you please be more explicit "how to use `std::nan()`" to tell whether `val` is NaN or not?

this logic is in order to pass [test_max()](https://github.com/pytorch/pytorch/blob/master/test/test_torch.py#L697) from `test_torch.py`, i was following [THTensor_(max)](https://github.com/pytorch/pytorch/blob/master/aten/src/TH/generic/THTensorMoreMath.cpp#L74) basically.

File: aten/src/ATen/native/cpu/TensorCompareKernel.cpp (hunk 2 above)
Comment (SsnL) [reply aten/src/ATen/native/cpu/TensorCompareKernel.cpp]: @mingfeima can you make this change? I think this would be good to go afterwards.

Comment (mingfeima) [reply aten/src/ATen/native/cpu/TensorCompareKernel.cpp]: @SsnL the macro is replaced by if statement now.

**Review by soumith — APPROVED** [review]

**Pull request closed without merging**